### PR TITLE
support chatglm in coati

### PR DESCRIPTION
**Pull request by yingliu-hpc**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 📝 What does this PR do?
support chatglm in coati

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: applications/Chat/coati/models/chatglm/chatglm_tokenizer.py [hunk 1]
```diff
@@ -0,0 +1,443 @@
+"""Tokenization classes for ChatGLM."""
```
Comment (TongLi3701) [inline applications/Chat/coati/models/chatglm/chatglm_tokenizer.py:1]: Add reference for this file.

File: requirements/requirements.txt [hunk 2]
```diff
@@ -10,4 +10,5 @@ contexttimer
 ninja
 torch>=1.12
 safetensors
+flash_attn==2.0.5
```
Comment (TongLi3701) [inline requirements/requirements.txt:13]: Do we need to add flash_attn here?

File: applications/Chat/tests/test_models.py [hunk 3]
```diff
@@ -129,12 +131,12 @@ def test_lora(lora_rank: int,
     # HACK: skip llama due to long execution time
     # lambda: (LlamaActor(), LlamaCritic(), LlamaRM()),
     lambda: (OPTActor(), OPTCritic(), OPTRM()),
+    lambda: (ChatGLMActor(), None, None),
 ])
 @torch.no_grad()
-def test_models(models_maker: Callable[[], Tuple[Actor, Critic, RewardModel]],
+def test_models(models_maker: Callable[[], Tuple[Actor, Critic, RewardModel, Actor]],
```
Comment (TongLi3701) [inline applications/Chat/tests/test_models.py:137]: Why do we need to add one more Actor here

File: applications/Chat/tests/test_models.py [hunk 4]
```diff
@@ -150,20 +152,30 @@ def test_models(models_maker: Callable[[], Tuple[Actor, Critic, RewardModel]],
...
 
     actor, critic, rm = models_maker()
+    if isinstance(actor, ChatGLMActor):
+        actor = actor.float()
+        tokenizer = ChatGLMTokenizer.from_pretrained( "THUDM/chatglm-6b", trust_remote_code=True)
+        chatglm_special_token = torch.tensor([tokenizer.gmask_token_id, tokenizer.bos_token_id]).repeat(batch_size, 1)
+        actor_input ={
+        "input_ids": torch.cat((torch.randint(0, 100, (batch_size, seq_len//2)), chatglm_special_token, torch.randint(0, 100, (batch_size, seq_len//2 - 2))), dim=1),
+        "attention_mask": torch.randint(0, 2, (batch_size, 1, seq_len, seq_len))
+    }
```
Comment (TongLi3701) [inline applications/Chat/tests/test_models.py:162]: What is this for?

In the main function, you only have bloom so this part will not be used. 

Also, you will need to assert some values or format or shapes but what do you want to test?

File: applications/Chat/tests/test_dataset.py [hunk 5]
```diff
@@ -213,6 +214,20 @@ def test_sft_dataset(model: str,
...
+            input_ids = sft_dataset[i]["input_ids"]
+            labels = sft_dataset[i]["labels"]
+            attention_mask = sft_dataset[i]["attention_mask"].to(torch.bool)
+            assert input_ids.shape == labels.shape == torch.Size([max_length])
+            assert attention_mask.shape[1] == attention_mask.shape[2] == max_length
+            
+            ignore_mask = labels == IGNORE_INDEX
+            assert input_ids.masked_select(torch.logical_not(ignore_mask))[0] == tokenizer.bos_token_id
+            return
+
```
Comment (TongLi3701) [inline applications/Chat/tests/test_dataset.py:230]: As you have implemented a special pre-processing method for chatglm. Your tests should test if the output follows the correct process and generate the output correctly.

File: applications/Chat/coati/dataset/sft_dataset.py [hunk 6]
```diff
@@ -71,6 +71,29 @@ def _preprocess(sources: Sequence[str],
...
 
 
+def _preprocess_chatglm(sources: Sequence[str],
+                targets: Sequence[str],
+                tokenizer: PreTrainedTokenizer,
+                max_length: int,
+                ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+    """Preprocess the data by tokenizing."""
+    sequences = [s + "[gMASK]" + tokenizer.bos_token + t for s, t in zip(sources, targets)]
+    sequences_token = tokenizer(sequences,
```
Comment (TongLi3701) [inline applications/Chat/coati/dataset/sft_dataset.py:81]: Why is it necessary to add [gMask] here, do we have any reference?

**Review by TongLi3701 — CHANGES_REQUESTED** [review]
Thanks Ying.

Some comments.

File: applications/Chat/coati/models/chatglm/chatglm_tokenizer.py (hunk 1 above)
Comment (yingliu-hpc) [reply applications/Chat/coati/models/chatglm/chatglm_tokenizer.py:4]: added

File: requirements/requirements.txt (hunk 2 above)
Comment (yingliu-hpc) [reply requirements/requirements.txt]: removed

File: applications/Chat/tests/test_models.py (hunk 3 above)
Comment (yingliu-hpc) [reply applications/Chat/tests/test_models.py]: fixed

File: applications/Chat/tests/test_models.py (hunk 4 above)
Comment (yingliu-hpc) [reply applications/Chat/tests/test_models.py:162]: This is the unit test for models in coati. ChatGLM is added in the pytest parameter list, which will be executed in ci.

File: applications/Chat/tests/test_dataset.py (hunk 5 above)
Comment (yingliu-hpc) [reply applications/Chat/tests/test_dataset.py]: Added

File: applications/Chat/coati/dataset/sft_dataset.py (hunk 6 above)
Comment (yingliu-hpc) [reply applications/Chat/coati/dataset/sft_dataset.py]: gMASK is the special token in ChatGLM

File: applications/Chat/coati/dataset/sft_dataset.py [hunk 7]
```diff
@@ -71,6 +71,38 @@ def _preprocess(sources: Sequence[str],
...
+    None for attention mask, ChatGLM will calculate attention mask according to input ids
+    """
+  
+    labels = []
+    input_ids = []
+    for source, target in zip(sources, targets):
+        source_id = tokenizer.encode(text=source, add_special_tokens=False)
+        target_id = tokenizer.encode(text=target, add_special_tokens=False)
+        if len(source_id) + len(target_id) >= max_length - 2:
+            source_id = source_id[ : max(0, max_length - len(target_id) - 2)]
```
Comment (TongLi3701) [inline applications/Chat/coati/dataset/sft_dataset.py:90]: You should remove extra from the left side.

File: applications/Chat/coati/dataset/sft_dataset.py [hunk 8]
```diff
@@ -71,6 +71,38 @@ def _preprocess(sources: Sequence[str],
...
+        if len(target_id) > max_length - len(source_id) - 2:
+            target_id = target_id[: max(0, max_length - len(source_id) - 3)] + [target_id[-1]]
```
Comment (TongLi3701) [inline applications/Chat/coati/dataset/sft_dataset.py:92]: A bit confused about this part. 

It can be sample, you can first build the sequence and then remove the extra from the left.

File: applications/Chat/coati/dataset/sft_dataset.py (hunk 7 above)
Comment (yingliu-hpc) [reply applications/Chat/coati/dataset/sft_dataset.py]: Fixed

File: applications/Chat/coati/dataset/sft_dataset.py (hunk 8 above)
Comment (yingliu-hpc) [reply applications/Chat/coati/dataset/sft_dataset.py]: Fixed

**Review by chengeharrison — APPROVED** [review]
Thanks Liu ying.

**Review by TongLi3701 — APPROVED** [review]
Thanks Ying.

File: applications/Chat/coati/dataset/sft_dataset.py [hunk 9]
```diff
@@ -137,14 +176,22 @@ def __init__(self,
...
-        return dict(input_ids=self.input_ids[idx],
-                    labels=self.labels[idx],
-                    attention_mask=self.attention_mask[idx])
+        if self.attention_mask is not None:
+            return dict(input_ids=self.input_ids[idx],
+                        labels=self.labels[idx],
+                        attention_mask=self.attention_mask[idx])
+        else:
+            return dict(input_ids=self.input_ids[idx],
+                        labels=self.labels[idx])
```
[the remark this answers is not in the document]
Comment (yingliu-hpc) [reply applications/Chat/coati/dataset/sft_dataset.py:201]: Actually the ChatGLM model calculates attention mask in the ChatGLMForConditionalGeneration class. Besides,  based on the ptuning of ChatGLM repo, the attention mask is also not provided. Here we followed them :)
